File: leetcode/0207-course-schedule/solution.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        graph = defaultdict(list)
        for u , v in prerequisites:
            graph[v].append(u)
        unvisit = 0
        visit   =1
        visited = 2
        state = [unvisit] * numCourses
        def dfs(node):
            s = state[node]
            if s == visited: return True
            if s == visit: return False
            state[node] = visit
            for nei in graph[node]:
                if not dfs(nei): 
                    return False
            state[node] = visited
            return True
        for i in range(numCourses):
            if not dfs(i): return False
        return True
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        

File: leetcode/0207-course-schedule/solution.ts
function canFinish(numCourses: number, prerequisites: number[][]): boolean {
    let graph = {}
    for(let i=0;i<numCourses;i++){
        graph[i] = []
    }
    for(let [u,v] of prerequisites){
        graph[u].push(v)
    }
    let visited = new Set()
    let review = new Set()
    function dfs(node){
        if(review.has(node)){return true}
        if(visited.has(node)){return false}
        visited.add(node)
        review.add(node)
        for(let n of graph[node]){
            if(dfs(n)){
                return true
            }
        }
        review.delete(node)
        return false

    }
    for(let node of Object.keys(graph)){
        if(dfs(node)){
            return false
        }
    }
    console.log(graph)
    return true
};
